### DISCUSSION TITLE
Get/set multiple instrument properties at once

### DISCUSSION THREAD

**Discussion by samcondon4** [pymeasure/pymeasure#477] [Ideas]
Say that I have an instance called sr830 of the pymeasure instrument SR830 class. The SR830 class has several properties created via Instrument.control() and Instrument.measurement(), like phase, frequency, sensitivity, time_constant, etc. What I would like to be able to do is have an additional property defined in the Instrument base class called, say "properties" so that I can run:

        sr830.properties

And all of the properties created with Instrument.control() and Instrument.measurement() will be returned in a dictionary like:

        {"phase": (float) sr830 phase, "frequency": (float) sr830 frequency, "sensitivity": (float) sr830 sensitivity, "time constant": 
          (float) sr830 time constant value, etc....}

Then, I would also like to be able to set multiple properties in one line with something like:

        sr830.properties = {"phase": (float) phase value, "sensitivity": (float) sensitivity value, etc...}

I can think of a few ways of accomplishing this through a few modifications to the Instrument base class. Is this something that people think would be useful for pymeasure? Any thoughts on the best way(s) to accomplish this?

**Comment by CasperSchippers** [comment]
I think this can be a useful feature.
Maybe it is possible to have list attribute in the instrument class where the name of the attributes is stored. This list can be populated by the `control`, `measurement`, and `setting` property factories.
Here, you might to have to watch out that the list doesn't become too long for some instruments, with properties that might not always be valid to be queried; i think some instruments have properties that are only valid in some instances and can generate an error when queried otherwise.
Also, some instruments have properties that are not made with a property factory, but as an "ordinary" property; hence, if the property-list is automatisch populated in the property-factories, these will not be included.
These are some thing to think about for this feature. That being said, I think it could indeed be quite a useful feature.

**Reply by bilderbuchi** [reply]
I have to disagree here. You already raise the salient point
> Here, you might to have to watch out that the list doesn't become too long for some instruments, with properties that might not always be valid to be queried; i think some instruments have properties that are only valid in some instances and can generate an error when queried otherwise.

This is definitely the case, and will not make such a list generally useful. I have done this before with a Keithley2400 - queried all properties at startup to populate GUI elements. Aside from the fact that querying all the properties takes quite some time, let's say, the Keithley did not react favorably -- beeping and error messages (wrong mode, blabla not allowed, etc) were frequent, I think less than half the properties gave useful values back.

Having said that, if you (OP) still want to do this, you don't need modifications of the `Instrument` class at all to achieve this. Make a method that loops through the attributes of your instrument class, selects those who are properties, filter on those which have `fget` defined, query, and use the attribute name and returned value to populate your dictionary. 
Something like this
```
for attr in dir(obj):
    if isinstance(getattr(type(obj), attr, None), property):
        print(attr, " is a property")
```
should be a good start.

Setting multiple properties is even easier, loop through the passed dict and use setattr.

**Reply by samcondon4** [reply]
I agree that having a list of ALL instrument properties would not be helpful as not all instrument properties can be queried while others are not useful. Though I do think that it would be quite nice to be able to query/set some subset of the instrument's properties in a one-liner with a dictionary. One solution is to introduce two new properties to the instrument base class. The first called something like "quick_properties_list" which is a list of all properties that the user wants to be able to query/set in one go, while the second property is called something like "quick_properties." 

Taking my SR830 example again, say that the only properties that I want to be able to query/set with a dictionary are the phase, sensitivity, and time constant. Then using the aforementioned two properties I would be able to do:

sr830 = SR830(<GPIB address)
sr830.quick_properties_list = ["phase", "sensitivity", "time_constant"]
print(sr830.quick_properties)

Which would print out something like:

{'phase': 111.36, 'sensitivity': 0.5, 'time_constant': 1}

Then, I could set the phase, sensitivity, and time constant in one go with:

sr830.quick_properites = {"phase": 45, :"sensitivity": 0.05, "time_constant": 0.1}

This technique does not keep a running list of EVERY property created, it simply creates a convenience property that the user can customize according to their needs. Some type checking could be implemented as well so that the user can't set a property in the quick_properties_list that cannot be queried.
